p2p/lightchain.py

```diff
@@ -210,19 +224,62 @@ def peer_block_height(peer: LESPeer) -> int:
             return
 
         start_block = await self.get_sync_start_block(peer, head_info)
-        while start_block < head_info.block_number:
+        while start_block < head_info.block_number and peer.connected:
             try:
                 # We should use "start_block + 1" here, but we always re-fetch the last synced
                 # block to work around https://github.com/ethereum/go-ethereum/issues/15447
                 batch = await self.fetch_headers(start_block, peer)
             except TooManyTimeouts:
                 raise LESAnnouncementProcessingError(
                     "Too many timeouts when fetching headers from {}".format(peer))
-            for header in batch:
-                await self.headerdb.coro_persist_header(header)
-                start_block = header.block_number
+            new_tip = await self._import_headers_from_peer(batch, peer)
+            if new_tip is not None:
+                start_block = new_tip
             self.logger.info("synced headers up to #%s", start_block)
 
+    async def _import_headers_from_peer(self, headers, peer):
+        """
+        :return: the block number of the new tip after importing the header
+        """
+        new_tip = None
+        for header in headers:
+            try:
+                await self._validate_header(header)
+            except ValidationError as exc:
+                self.logger.info(
+                    "Dropping Peer %s for sending invalid header %s: %s",
+                    peer,
+                    header,
+                    exc,
+                )
```

Comment:
I think it'd be better if you raised a `LESAnnouncementProcessingError` here, including the DisconnectReason there. That will cause a warning to be logged and the peer to be disconnected. Doing this you also won't need the `peer.connected` check and the `new_tip is not None` checks in `process_announcement()`